### PR TITLE
Implementing Proposal 4 - Static Compute Stream

### PR DESCRIPTION
**Pull request by obaidott**
Implementations.. Proposal 4 - To each his/her own

- [x] .toStreamFromCompute
- [x] .toStreamFromPropery(obs, propName )
- [x] .toStreamFromEvent(obs, event)
- [x] .toStreamFromEvent(obs, propName, event)
- [x] .toStream

### CODE REVIEW

File: can-compute-stream.js [hunk 1]
```diff
@@ -1,80 +1,163 @@
 var compute = require('can-compute');
-var define = require('can-define');
 var Kefir = require('kefir');
 var makeArray = require("can-util/js/make-array/make-array");
 var isArray = require("can-util/js/is-array/is-array");
 var assign = require("can-util/js/assign/assign");
+var define = require('can-define');
```
Comment (justinbmeyer) [inline can-compute-stream.js]: Why would can-define be needed?  You might want to add jshint (you can see how other projects have it).

File: can-compute-stream.js [hunk 2]
```diff
@@ -1,80 +1,163 @@
...
 			compute.off('change', changeHandler);
 		};
 	});
 };
 
+// Converts all arguments passed to streams, and resolves the resulting
+// streams to a single stream
+//   Assumes all arguments are computes
+//   Last argument is optionally a function
 var computeStream = function () {
```
Comment (justinbmeyer) [inline can-compute-stream.js]: I think this should be renamed something else.  Why is this function the export?  I'd expect an object to be exported.

File: can-compute-stream.js [hunk 3]
```diff
@@ -1,80 +1,163 @@
...
 	var streams = computes.map(singleComputeToStream);
-
+	//streams.push(Kefir.merge(computes));
+	// Resolves all of the streams to a single stream
 	return evaluator.apply(this, streams);
 };
 
+// A variation of `computeStream`, but pipes the values of the
+// resolved stream back into a compute
 computeStream.asCompute = function () {
```
Comment (justinbmeyer) [inline can-compute-stream.js]: Is `asCompute` used anymore?  If yes, it should probably be renamed.

File: can-compute-stream.js [hunk 4]
```diff
@@ -1,80 +1,163 @@
...
+	return computeStream.apply(this, arguments);
+};
+
+computeStream.toStreamFromProperty = function(obs, propName ) {
+	var compute = compute(obs, propName);
+	return computeStream.apply(this, compute);
+};
+
+computeStream.toStreamFromEvent = function(obs, eventName) {
+	var compute = compute(obs, eventName);
```
Comment (justinbmeyer) [inline can-compute-stream.js]: this isn't working.

File: can-compute-stream.js [hunk 5]
```diff
@@ -1,80 +1,163 @@
...
+	return computeStream.apply(this, compute);
+};
+
+computeStream.toComputeFromEvent = function(obs, propName, eventName) {
```
Comment (justinbmeyer) [inline can-compute-stream.js]: this wont work if propName changes.  Also, this won't work with deep propNames.

File: can-compute-stream.js [hunk 6]
```diff
@@ -1,80 +1,163 @@
...
+	var localCompute = computeStream.toComputeFromEvent(obs, propName, eventName);
+	return computeStream(localCompute);
+};
+
+
 define.behaviors.push('stream');
 
 var oldExtensions = define.extensions;
-
 define.extensions = function (objPrototype, prop, definition) {
```
Comment (justinbmeyer) [inline can-compute-stream.js]: I thought this was going to be in a different repo.

File: can-compute-stream.js (hunk 5 above)
Comment (justinbmeyer) [inline can-compute-stream.js]: can't do `foo.bar`

**Comment by obaidott** [comment]
@justinbmeyer - if you get chance, can you review the PR again?

File: can-compute-stream.js [hunk 7]
```diff
@@ -32,56 +43,83 @@ var computeStream = function () {
...
+	var eventName, propName;
+
 
-	return valueCompute;
+	if(arguments.length === 2) {
+		//.toStreamFromEvent(obs, event);
+
+		eventName = arguments[1];
+		
+        return Kefir.stream(function (emitter) {
```
Comment (justinbmeyer) [inline can-compute-stream.js]: can you make sure your indentation is uniform.  Use tabs.  THanks!

File: can-compute-stream.js [hunk 8]
```diff
@@ -32,56 +43,83 @@ var computeStream = function () {
...
-				return computeStream.asCompute.apply(this, computes);
-			}
-		}, define.types.compute);
-	} else {
-		return oldExtensions.apply(this, arguments);
-	}
+/*
+ * Takes multiple streams and returns a single stream
+ */
+computeStream.toSingleStream = function() {
```
Comment (justinbmeyer) [inline can-compute-stream.js]: is this needed?

File: can-compute-stream.js [hunk 9]
```diff
@@ -1,26 +1,37 @@
...
 };
 
-var computeStream = function () {
+
+/*
+ * Converts all arguments passed into a single stream and resolves the resulting
+ * streams into a single stream. Assumes all arguments are computes and last
+ * argument is optionally a function.
+ */
+computeStream.toStreamFromCompute = function () {
```
Comment (justinbmeyer) [inline can-compute-stream.js]: does this need to take multiple arguments? 

You don't need to use `hasOwnProperty` for isComputed.  Just read `.isComputed`

File: can-compute-stream.js (hunk 8 above)
Comment (obaidott) [inline can-compute-stream.js]: I am using this in a test case that merges multiple streams into a single stream.

File: can-stream.js [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+};
+
+/*
+ * A single API to handle most of the cases
+ */
+canStream.toStream = function() {
+
+	if(arguments.length === 1) {
+		//we expect it to be a compute:
+		return canStream.toStreamFromCompute(arguments[0]); //#1
```
Comment (justinbmeyer) [inline can-stream.js]:
```
// toStream( compute )
```

File: can-stream.js [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+			else {
+				return canStream.toStreamFromEvent(obs, eventNameOrPropName); //#3a
+			}
+		}
+		else {
+			var splitEventNameAndProperty = eventNameOrPropName.split(" ");
+			return canStream.toStreamFromEvent(obs, splitEventNameAndProperty[0].slice(1), splitEventNameAndProperty[1]);  //#3b
+		}
+	}
+	return false; //we reached the end for some reason without creating any stream.
```
Comment (justinbmeyer) [inline can-stream.js]: return undefined

File: can-stream.md [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+  @param {string} property name
+
+  @return {Stream} A [Kefir](https://rpominov.github.io/kefir/) stream.
+
+@signature `canStream.toStreamFromEvent(obs, propName)`
+
+Creates a stream based on property value change on observable
+
+  @param {observable} An observable object
+  @par am {string} property name
```
Comment (justinbmeyer) [inline can-stream.md]: @param

**Pull request merged**